plugins/core/IO/qLASIO/src/LasWaveformLoader.cpp

```diff
@@ -0,0 +1,244 @@
+                                                                  unsigned int numVlrs)
+{
+    ccPointCloud::FWFDescriptorSet descriptors;
+
+    for (size_t i{0}; i < numVlrs; ++i)
+    {
+        const laszip_vlr_struct *vlr = &vlrs[i];
+        if (vlr && strcmp(vlr->user_id, "LASF_Spec") == 0 && 99 < vlr->record_id && vlr->record_id < 355)
+        {
+            WaveformDescriptor descriptor;
+            if (!ParseWavepacketDescriptorVlr(*vlr, descriptor))
+            {
+                ccLog::Warning("[LAS] Invalid Descriptor VLR");
+            }
+            else
+            {
+                descriptors.insert(vlr->record_id - 99, descriptor);
+            }
+        }
+    }
+    return descriptors;
+}
+
+LasWaveformLoader::LasWaveformLoader(const laszip_header_struct &laszipHeader,
+                                     const QString &lasFilename,
+                                     ccPointCloud &pointCloud)
+    : isPointFormatExtended(laszipHeader.point_data_format >= 6)
+{
+    descriptors =
+        parseWaveformDescriptorVlrs(laszipHeader.vlrs, laszipHeader.number_of_variable_length_records);
+    ccLog::Print("[LAS] %d Waveform Packet Descriptor VLRs found", descriptors.size());
+
+    QFile fwfDataSource;
+    if (laszipHeader.start_of_waveform_data_packet_record != 0)
+    {
+        ccLog::Print("[LAS] Waveform data is located within the las file");
+        fwfDataSource.setFileName(lasFilename);
+        if (!fwfDataSource.open(QFile::ReadOnly))
+        {
+            ccLog::Warning(
+                QString("[LAS] Failed to re open the las file: %1").arg(fwfDataSource.errorString()));
+            return;
+        }
+
+        if (!fwfDataSource.seek(laszipHeader.start_of_waveform_data_packet_record))
+        {
+            ccLog::Warning(QString("[LAS] Failed to find the associated waveform data packets header"));
+            return;
+        }
+
+        QDataStream stream(&fwfDataSource);
+        EvlrHeader evlrHeader;
+        stream >> evlrHeader;
+        if (stream.status() == QDataStream::Status::ReadPastEnd)
+        {
+
+            ccLog::Warning(QString("[LAS] Failed to read the associated waveform data packets"));
+            return;
+        }
+
+        if (!evlrHeader.isWaveFormDataPackets())
+        {
+            ccLog::Warning("[LAS] Invalid waveform EVLR");
+            return;
+        }
+        fwfDataCount = evlrHeader.recordLength;
+        fwfDataOffset = 0;
+        if (fwfDataCount == 0)
+        {
+            ccLog::Warning(QString("[LAS] Invalid waveform data packet size (0). We'll load all the "
+                                   "remaining part of the file!"));
+            fwfDataCount = fwfDataSource.size() - fwfDataSource.pos();
+        }
+    }
+    else if (laszipHeader.global_encoding & 4)
+    {
+        QFileInfo info(lasFilename);
+        QString wdpFilename = info.path() + "/" + info.completeBaseName() + ".wdp";
```

Comment:
I fixed a bug here (completeBaseName should be used instead of baseName for files with '.' inside their name).

Have you re-used any of the code in the LASFWFIO plugin, or have you really redone everything from scratch?

Reply:
From what I remember (this plugin was actually written 1-2 years ago) most of it is just redone from scratch with the LAS spec on the 2nd screen. 

I did look at how it was done in the FWF plugin to get started on how to put the fwf data and sometimes check the ideas were the same.

Reply:
What about the PDAL plugin GUI? And all the features that come with it? (being able to split the classification field, tiling the cloud, etc.)? Do you plan to reuse the GUI at least? (or at least part of it?)

Reply:
The GUI is also from scratch

- For the loading file part, the fields from classification are already split by default (with no possibility of re-merging them) as in the las standard, you don't have the choice the classification is on 5 bits, the 3 following bits are there own 1-bit fields (for LAS version < 1.4)

- For the writing, it is also from scratch and it provides more features, like
   -  selecting the version you wish to save the file into
   - which cc scalar field you want to map to LAS field with warnings If data loss due to precision would occur. 
   - And later I plan to add  the ability to add cc scalarfields as LAS extra scalar fields (as of now, cc scarlar fields are saved as LAS extra fields only if they were LAS extra scalar field in the first place)

- I didn't add tiling, I always found it weird that the LAS filter was the only one that had this option.
 But if its really necessary it can be re-added

Reply:
Let's see how this is received then ;)